### ISSUE TITLE
New state for modules

### ISSUE THREAD

**Issue by pascalgrimaud** [jhipster/jhipster-lite#10934]
Today, we have 2 states for modules:
- available, displayed in the UI
- disabled, which are hidden, specially because it's not maintained

I want to add a new state:
- beta
- wip
- or another name

These modules will be displayed with a warning or another color (any suggestions are welcomed)

So we can start new modules and get some feedbacks, and totally rework if needed, without worrying about the use

What do you think?
Good or bad idea?

**Comment by renanfranca** [comment]
@pascalgrimaud : I believe it is an excellent idea!

I have often felt the need to release something with beta status at places where I've worked, but it was never allowed because they preferred to keep testing internally until the functionality was complete.

> I want to add a new status:
beta
wip
or another name

I think the term "beta" encourages testing more than "wip" (which suggests it might not even be usable).

> These modules will be displayed with a warning or another color (any suggestions are welcomed)

Regarding the interface, I think a light orange color and a small label reading "beta" directly on the module box would look nice.

If the idea 💡 is approved, I volunteer for this task 🤩

**Comment by renanfranca** [comment]
@pascalgrimaud: I am looking for the next issue to work on. Can I start working on this feature?

**Comment by pascalgrimaud** [comment]
Yes you can, @renanfranca but I have more ideas on that. I don't want to limit to 2 states (availabled / disabled).

We should use something like a priority (a value between 1 to 5 for example). Alternatively, we could use a ranking system like in manga, such as Rank S, Rank A, Rank B, etc., to determine which modules to display.

So we can choose to display what we want:
- all modules
- only modules with rank S
- only modules with rank S and A
- etc...

As you can see in this screen:

![image](https://github.com/user-attachments/assets/71c45af9-3025-4759-8a97-ad9caa637cd2)

There are too many modules now, and some are not used frequently. We should have the ability to decide whether to display them or not.

Maven or Spring Boot must be Rank S
htmx-webjars (sorry) should be Rank B or C, or lower

**Comment by pascalgrimaud** [comment]
cc @murdos : what do you think?

**Comment by pascalgrimaud** [comment]
Let's start by small steps:
- back: add new information which is related to RANK_S, RANK_A, etc
- back: each module can have a Rank, but by default, it would be RANK_D (the lowest)

**Comment by renanfranca** [comment]
> Let's start by small steps:
> - back: add new information which is related to RANK_S, RANK_A, etc
> - back: each module can have a Rank, but by default, it would be RANK_D (the lowest)

Thanks 👍.  Should I implement the beta state, too?

**Comment by pascalgrimaud** [comment]
No need, beta state will be RANK_D

**Comment by renanfranca** [comment]
> But we need something to filter or to highlight the main and important modules. Currently, I use ctrl+f when searching postgres, and it's not really nice :-D

@pascalgrimaud : I've got two ideas that might help:

1 - Add an extended JHipster Lite module that lets you hide modules based on tags.

2 - Or maybe we could add an option on the landscape screen to hide or highlight modules by tags. That setting could be saved in local storage so you don't have to reselect it every time you hit F5.

WDYT?

**Comment by renanfranca** [comment]
> It could be a search input, like in patches page
> 
> ![image](https://private-user-images.githubusercontent.com/9156882/374972545-ffc70ae3-ac83-42a1-8b60-e1b912fb3432.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg0ODAxMzIsIm5iZiI6MTcyODQ3OTgzMiwicGF0aCI6Ii85MTU2ODgyLzM3NDk3MjU0NS1mZmM3MGFlMy1hYzgzLTQyYTEtOGI2MC1lMWI5MTJmYjM0MzIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMDlUMTMxNzEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODZjYWY5YjMxMjkxNDZlMGIyNDIxZmIzMWZkZDViMWZlMGZlOTZhM2MyYzMyMjY1ZjhlYzY4MWQzMjIxYWQ1ZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.2oMjjjsiF8eCQcwyDNWHvBC7Zu5G4KH8XsL0iRyMoTk)

![image](https://github.com/user-attachments/assets/74c1ea3f-0e09-49f4-9a27-844b7e673721)

In the picture above, the search bar automatically positions the module in the right place. So, should the input text center the module on the screen? Or should it have a different behavior? My concern is placing the search input text in a way that doesn't interfere with the module's visualization.

**Comment by renanfranca** [comment]
@pascalgrimaud : I created a quick demo of how I think the search bar should work (ignore the search button, which wasn't needed). WDYT?


https://github.com/user-attachments/assets/8a067c9f-61bf-4380-ab51-c8995fcd1205

**Comment by pascalgrimaud** [comment]
Great @renanfranca !
It's nice to have but there should be a dedicated ticket for that ;)

**Comment by renanfranca** [comment]
> Great @renanfranca ! It's nice to have but there should be a dedicated ticket for that ;)

Thanks! I will create a new issue for that ;)

**Comment by DamnClin** [comment]
For me, the idea of module rank is good. I think ranks should go on landscape screen and we can remove replay.

An idea to rank modules:

* **D**: Not really usable as is, unless you have good expertise on the subject (ex: custom-jhlite);
* **C**: No known usage on production product (ex: vue-core, the current version);
* **B**: One declared usage on production product (ex: kipe-authorization);
* **A**: Multiple declared usages on production product, by multiple person, on various projects and demonstrated usage on a talk, book or blog post (ex: java-base);
* **S**: A and recognized to add some features that are not available otherwise. Features must be really unique and this must be recognized by at least 10 person with a vote on any social network (github beeing a social network). With those criteria spring-boot-cucumber-mvc is a good candidate (missing votes).

Here production product is a product that's not a toy project and actually meant for end users.

I also agree that we should rework the landscape. I build that thing on a 34" monitor with way less module than we have now. The whole thing became totally irrelevant. I think this should be done in a mob session and this is probably a prerequisite for that ranking feature.

**Comment by pascalgrimaud** [comment]
So it's time to start this, @renanfranca do you want to do it ?

**Comment by renanfranca** [comment]
> So it's time to start this, @renanfranca do you want to do it ?

@pascalgrimaud : Yes, absolutely! What exactly is the feature I should implement?

Thanks! 😁

**Comment by renanfranca** [comment]
> For me, the idea of module rank is good. I think ranks should go on landscape screen and we can remove replay.
> 
> An idea to rank modules:
> 
> * **D**: Not really usable as is, unless you have good expertise on the subject (ex: custom-jhlite);
> * **C**: No known usage on production product (ex: vue-core, the current version);
> * **B**: One declared usage on production product (ex: kipe-authorization);
> * **A**: Multiple declared usages on production product, by multiple person, on various projects and demonstrated usage on a talk, book or blog post (ex: java-base);
> * **S**: A and recognized to add some features that are not available otherwise. Features must be really unique and this must be recognized by at least 10 person with a vote on any social network (github beeing a social network). With those criteria spring-boot-cucumber-mvc is a good candidate (missing votes).
> 
> Here production product is a product that's not a toy project and actually meant for end users.
> 
> I also agree that we should rework the landscape. I build that thing on a 34" monitor with way less module than we have now. The whole thing became totally irrelevant. I think this should be done in a mob session and this is probably a prerequisite for that ranking feature.

So the feature is to make it possible to add Ranks in the module configuration (the same place where we add module dependencies and set features) and filter what should be shown through application.yml?

Cc: @pascalgrimaud

**Comment by pascalgrimaud** [comment]
@renanfranca : no, 1 module should have only one Rank. I think the rank can be optional, and if it's the case, it will be rank D by default, it will be easier for a 1st version -> no need to add a Rank to all existing modules

**Comment by renanfranca** [comment]
> @renanfranca : no, 1 module should have only one Rank. I think the rank can be optional, and if it's the case, it will be rank D by default, it will be easier for a 1st version -> no need to add a Rank to all existing modules

Thanks, nice!

@pascalgrimaud : Should I add a configuration option to hide the rank, similar to the existing configuration for tags in `application.yml`, in this 1st version?

**Comment by pascalgrimaud** [comment]
Why do you want to hide the rank ? Anywayt, it won't be used until the front is improved.

The first version / first PR could be related to domain only, not related to configuration:
- the module has a new field Rank, which is Optional
- you can build a Rank, with 5 values: Rank S, Rank A, Rank B, Rank C, Rank D

Then later, we can work on mapper and Rest DTO

**Comment by renanfranca** [comment]
> Why do you want to hide the rank?

I read the issue again and found that you mentioned [that before](https://github.com/jhipster/jhipster-lite/issues/10934#issuecomment-2399999505):

> There are too many modules now, and some are not used frequently. We should have the ability to decide whether to display them or not.

That's why I suggested the hide rank configuration option :)

> Anyway, it won't be used until the front is improved.

> The first version/first PR could be related to the domain only, not related to configuration:
- The module has a new field Rank, which is Optional.
- You can build a Rank with 5 values: Rank S, Rank A, Rank B, Rank C, Rank D.

Thanks for the detailed description 😀

> Then later, we can work on the mapper and Rest DTO

Nice 👍

**Comment by pascalgrimaud** [comment]
Adding a bounty on this beautiful feature.
I'll test it soon @renanfranca 

Thanks for your patience

**Comment by renanfranca** [comment]
> Adding a bounty on this beautiful feature.
> I'll test it soon @renanfranca 
> 
> Thanks for your patience

Thank you very much! I loved the result  😃🙌

**Comment by renanfranca** [comment]
@pascalgrimaud: Bounty Claimed  https://opencollective.com/generator-jhipster/expenses/243570. Thanks 🙌💙

**Comment by pascalgrimaud** [comment]
@renanfranca : approved, thanks for this wonderful feature

**Comment by renanfranca** [comment]
> @renanfranca : approved, thanks for this wonderful feature

You are welcome! I am really proud of the result! 💖
